src/thumbnailer_slope_optim.cc

```diff
@@ -27,13 +27,26 @@ Thumbnailer::Status Thumbnailer::GenerateAnimationSlopeOptim(
   CHECK_THUMBNAILER_STATUS(LossyEncodeSlopeOptim(webp_data));
   CHECK_THUMBNAILER_STATUS(NearLosslessEqual(webp_data));
 
+  // If there is no frame encoded in near-lossless, use
+  // GenerateAnimationEqualQuality() to generate animation.
```

Comment:
I do not understand why the SlopeOptim is discarded if there is no near-lossless frame. It is a lossy method, right?

Reply:
If there's no near-lossless frame, calling `GenerateAnimationEqualQuality()` is to use lossy encoding with the same quality and also to save time rather than calling `LossyEncodeNoSlopeOptim` and then `ExtraLossyEncode`.

Reply:
So we have:
```c++
LossyEncodeSlopeOptim();
NearLosslessEqual();
if (at_least_one_frame_is_near_lossless) {
  LossyEncodeNoSlopeOptim();
  ExtraLossyEncode();
} else {
  GenerateAnimationEqualQuality();
}
```
Are you saying `GenerateAnimationEqualQuality()` has the same output as `LossyEncodeNoSlopeOptim()+ExtraLossyEncode()` only if there is no near-lossless?

Reply:
Yes, that's right.

Reply:
But the `encoding_order` stuff never happens in `GenerateAnimationEqualQuality()`. It happens in `ExtraLossyEncode()` only if `NearLosslessEqual()` frees some budget?

Why not stop at `LossyEncodeSlopeOptim()` when there is no near-lossless then? Does it outperform `GenerateAnimationEqualQuality()` only when there is at least one near-lossless frame? I am bit confused.

Reply:
In `LossyEncodeSlopeOptim()`, the qualities of all frames are reduced a bit to have more budget for `NearLosslessEqual()`, so if there's no near-lossless frame after using `NearLosslessEqual()`, generating animation with  `GenerateAnimationEqualQuality()` is the best choice in my opinion. In fact, `ExtraLossyEncoding` is used to get the highest qualities that are as close as possible to the resulting quality of `GenerateAnimationEqualQuality()`.

Reply:
Where is that "reduced a bit" part in `LossyEncodeSlopeOptim()`? I do not see it. Also is not `LossyEncodeNoSlopeOptim()` supposed to fill any budget gap left?

If I understood correctly, `ExtraLossyEncode()` is doing the same as `GenerateAnimationEqualQuality()` but several times on a smaller and smaller subset of frames. Thus I have trouble seeing why/when the latter is better than the former, with or without near-lossless frames (which are excluded from the binary search in both functions right?).

Anyway if you saw better results keep the code like that, but I am afraid that defaulting to `GenerateAnimationEqualQuality()` whenever there is no near-lossless (which should be the case quite a lot because the majority of animations is camera footage not fit for near-lossless) will make the slope optim barely usable. Also if there is a lot of near-lossless frames, lossy optims are also less valuable. The convenient spot for slope optim would be the right amount of near-lossless and lossy frames, which seems often unlikely.

Reply:
`LossyEncodeSlopeOptim()` move the quality to the left in cases the RD-curves are flat, so that it reduces the animation size. 

`LossyEncodeNoSlopeOptim()` can fill the unused budget but it is not really efficiently (as I mentioned before) so that's why I implemented `ExtraLossyEncode()`.

It is true that `ExtraLossyEncode()` does the same job as `GenerateAnimationEqualQuality()` but several times. So I set default to use `GenerateAnimationEqualQuality()` whenever there is no near-lossless frame just to avoid using `LossyEncodeNoSlopeOptim()` and then `ExtraLossyEncode()`. I tested and it seems like both methods have the same the results whenever there is no near-lossless frame, so for now I think I gonna remove `GenerateAnimationEqualQuality()` here in order to avoid misunderstanding, is it ok ?

Reply:
Yes, it is clearer, thanks!